File: engine/portfolio.py
"""Ranking and portfolio-level constraints.

`apply_portfolio_constraints` is ported line-for-line to web/src/lib/portfolio.ts and
both are checked against shared vectors (data/test_vectors/portfolio_constraints.json).
All money arithmetic is done in integer paise so Python and JavaScript agree exactly.
"""

from __future__ import annotations

from dataclasses import replace

from .config import ScanConfig
from .types import Candidate, OpenPosition, PortfolioBudget, RankStatus, Side


def paise(x: float) -> int:
    return int(round(x * 100))


def rank_candidates(cands: list[Candidate], cfg: ScanConfig) -> list[Candidate]:
    """Sort by raw score desc (symbol asc for deterministic ties); keep top N per side.
    Candidates without a sized plan (qty 0 or no plan) are not ranked."""
    out: list[Candidate] = []
    for side, top_n in ((Side.LONG, cfg.top_n_longs), (Side.SHORT, cfg.top_n_shorts)):
        pool = [c for c in cands if c.side is side]
        pool.sort(key=lambda c: (-c.score.raw, c.symbol))
        kept = 0
        for c in pool:
            if c.plan is None or c.plan.qty <= 0:
                out.append(replace(c, rank=None, rank_status=None, rank_reason="no sized plan"))
                continue
            if kept < top_n:
                kept += 1
                out.append(replace(c, rank=kept, rank_status=RankStatus.RANKED, rank_reason=None))
            else:
                out.append(replace(c, rank=None, rank_status=RankStatus.NOT_IN_TOP_N, rank_reason=f"outside top {top_n}"))
    return out


def apply_portfolio_constraints(
    ranked: list[Candidate], open_positions: list[OpenPosition], cfg: ScanConfig
) -> tuple[list[Candidate], PortfolioBudget]:
    """Walk ranked candidates in rank order and accept while all caps hold.

    Deferred candidates keep their rank but get a `deferred_*` status and reason.
    Order: longs then shorts by rank (the same order the TS port uses).
    """
    cap_p = paise(cfg.capital)
    gross_cap = cap_p * cfg.max_gross_exposure_pct / 100.0
    risk_cap = cap_p * cfg.max_portfolio_risk_pct / 100.0
    gross_cap_p = int(round(gross_cap))
    risk_cap_p = int(round(risk_cap))

    open_gross = sum(paise(p.position_value) for p in open_positions)
    open_risk = sum(paise(p.risk_amount) for p in open_positions)
    positions = len(open_positions)
    per_sector: dict[str, int] = {}
    held: set[str] = set()
    for p in open_positions:
        per_sector[p.sector or "UNKNOWN"] = per_sector.get(p.sector or "UNKNOWN", 0) + 1
        held.add(p.symbol)

    gross = open_gross
    risk = open_risk
    acc_gross = 0
    acc_risk = 0
    acc_n = 0

    def order_key(c: Candidate):
        return (0 if c.side is Side.LONG else 1, c.rank if c.rank is not None else 10**9, c.symbol)

    result: list[Candidate] = []
    for c in sorted(ranked, key=order_key):
        if c.rank_status is not RankStatus.RANKED or c.plan is None:
            result.append(c)
            continue
        sector = c.sector or "UNKNOWN"
        pv = paise(c.plan.position_value)
        ra = paise(c.plan.risk_amount)
        if c.symbol in held and c.side is Side.LONG and not cfg.allow_pyramiding:
            result.append(replace(c, rank_status=RankStatus.ALREADY_HELD, rank_reason="open position exists and allow_pyramiding=false"))
            continue
        if positions + 1 > cfg.max_concurrent_positions:
            result.append(replace(c, rank_status=RankStatus.DEFERRED_MAX_POSITIONS, rank_reason=f"max_concurrent_positions {cfg.max_concurrent_positions} reached"))
            continue
        if per_sector.get(sector, 0) + 1 > cfg.max_positions_per_sector:
            result.append(replace(c, rank_status=RankStatus.DEFERRED_SECTOR_CAP, rank_reason=f"max_positions_per_sector {cfg.max_positions_per_sector} reached for {sector}"))
            continue
        if gross + pv > gross_cap_p:
            result.append(replace(c, rank_status=RankStatus.DEFERRED_GROSS_EXPOSURE, rank_reason=f"gross exposure would exceed {cfg.max_gross_exposure_pct:g}% of capital"))
            continue
        if risk + ra > risk_cap_p:
            result.append(replace(c, rank_status=RankStatus.DEFERRED_RISK_BUDGET, rank_reason=f"portfolio risk would exceed {cfg.max_portfolio_risk_pct:g}% of capital"))
            continue
        gross += pv
        risk += ra
        acc_gross += pv
        acc_risk += ra
        acc_n += 1
        positions += 1
        per_sector[sector] = per_sector.get(sector, 0) + 1
        held.add(c.symbol)
        result.append(c)

    budget = PortfolioBudget(
        capital=cfg.capital,
        open_positions=len(open_positions),
        open_gross_value=open_gross / 100.0,
        open_risk_amount=open_risk / 100.0,
        accepted_positions=acc_n,
        accepted_gross_value=acc_gross / 100.0,
        accepted_risk_amount=acc_risk / 100.0,
        gross_cap=gross_cap_p / 100.0,
        risk_cap=risk_cap_p / 100.0,
        per_sector=dict(sorted(per_sector.items())),
    )
    # restore original order (by side then rank/symbol) for stable output
    result.sort(key=order_key)
    return result, budget

File: web/src/lib/portfolio.ts
// Line-for-line port of engine/portfolio.py::apply_portfolio_constraints.
// Verified against data/test_vectors/portfolio_constraints.json (see portfolio.test.ts).
// All money arithmetic is in integer paise so Python and JS agree exactly.

import type { RankStatus, Side } from "./types";

export interface ConstraintConfig {
  capital: number;
  max_portfolio_risk_pct: number;
  max_gross_exposure_pct: number;
  max_concurrent_positions: number;
  max_positions_per_sector: number;
  allow_pyramiding: boolean;
  top_n_longs: number;
  top_n_shorts: number;
}

export interface SlimCandidate {
  symbol: string;
  side: Side;
  sector: string | null;
  raw: number;
  qty: number;
  position_value: number;
  risk_amount: number;
}

export interface OpenPositionLike {
  symbol: string;
  side: Side;
  sector: string | null;
  position_value: number;
  risk_amount: number;
}

export interface RankedCandidate extends SlimCandidate {
  rank: number | null;
  rank_status: RankStatus | null;
  rank_reason: string | null;
}

export interface Budget {
  capital: number;
  open_positions: number;
  open_gross_value: number;
  open_risk_amount: number;
  accepted_positions: number;
  accepted_gross_value: number;
  accepted_risk_amount: number;
  gross_cap: number;
  risk_cap: number;
  per_sector: Record<string, number>;
}

export function paise(x: number): number {
  return Math.round(x * 100);
}

function cmp(a: RankedCandidate, b: RankedCandidate): number {
  if (b.raw !== a.raw) return b.raw - a.raw;
  return a.symbol < b.symbol ? -1 : a.symbol > b.symbol ? 1 : 0;
}

export function rankCandidates(cands: SlimCandidate[], cfg: ConstraintConfig): RankedCandidate[] {
  const out: RankedCandidate[] = [];
  const sides: [Side, number][] = [["long", cfg.top_n_longs], ["short", cfg.top_n_shorts]];
  for (const [side, topN] of sides) {
    const pool = cands.filter((c) => c.side === side).map((c) => ({ ...c, rank: null as number | null, rank_status: null as RankStatus | null, rank_reason: null as string | null }));
    pool.sort(cmp);
    let kept = 0;
    for (const c of pool) {
      if (c.qty <= 0) {
        out.push({ ...c, rank: null, rank_status: null, rank_reason: "no sized plan" });
        continue;
      }
      if (kept < topN) {
        kept += 1;
        out.push({ ...c, rank: kept, rank_status: "ranked", rank_reason: null });
      } else {
        out.push({ ...c, rank: null, rank_status: "not_in_top_n", rank_reason: `outside top ${topN}` });
      }
    }
  }
  return out;
}

function orderKey(c: RankedCandidate): [number, number, string] {
  return [c.side === "long" ? 0 : 1, c.rank ?? 1e9, c.symbol];
}

function byOrder(a: RankedCandidate, b: RankedCandidate): number {
  const ka = orderKey(a), kb = orderKey(b);
  if (ka[0] !== kb[0]) return ka[0] - kb[0];
  if (ka[1] !== kb[1]) return ka[1] - kb[1];
  return ka[2] < kb[2] ? -1 : ka[2] > kb[2] ? 1 : 0;
}

export function applyPortfolioConstraints(
  ranked: RankedCandidate[],
  openPositions: OpenPositionLike[],
  cfg: ConstraintConfig,
): { results: RankedCandidate[]; budget: Budget } {
  const capP = paise(cfg.capital);
  const grossCapP = Math.round((capP * cfg.max_gross_exposure_pct) / 100.0);
  const riskCapP = Math.round((capP * cfg.max_portfolio_risk_pct) / 100.0);

  const openGross = openPositions.reduce((s, p) => s + paise(p.position_value), 0);
  const openRisk = openPositions.reduce((s, p) => s + paise(p.risk_amount), 0);
  let positions = openPositions.length;
  const perSector: Record<string, number> = {};
  const held = new Set<string>();
  for (const p of openPositions) {
    const sec = p.sector ?? "UNKNOWN";
    perSector[sec] = (perSector[sec] ?? 0) + 1;
    held.add(p.symbol);
  }

  let gross = openGross, risk = openRisk, accGross = 0, accRisk = 0, accN = 0;
  const result: RankedCandidate[] = [];
  for (const c of [...ranked].sort(byOrder)) {
    if (c.rank_status !== "ranked") {
      result.push(c);
      continue;
    }
    const sector = c.sector ?? "UNKNOWN";
    const pv = paise(c.position_value);
    const ra = paise(c.risk_amount);
    if (held.has(c.symbol) && c.side === "long" && !cfg.allow_pyramiding) {
      result.push({ ...c, rank_status: "already_held", rank_reason: "open position exists and allow_pyramiding=false" });
      continue;
    }
    if (positions + 1 > cfg.max_concurrent_positions) {
      result.push({ ...c, rank_status: "deferred_max_positions", rank_reason: `max_concurrent_positions ${cfg.max_concurrent_positions} reached` });
      continue;
    }
    if ((perSector[sector] ?? 0) + 1 > cfg.max_positions_per_sector) {
      result.push({ ...c, rank_status: "deferred_sector_cap", rank_reason: `max_positions_per_sector ${cfg.max_positions_per_sector} reached for ${sector}` });
      continue;
    }
    if (gross + pv > grossCapP) {
      result.push({ ...c, rank_status: "deferred_gross_exposure", rank_reason: `gross exposure would exceed ${cfg.max_gross_exposure_pct}% of capital` });
      continue;
    }
    if (risk + ra > riskCapP) {
      result.push({ ...c, rank_status: "deferred_risk_budget", rank_reason: `portfolio risk would exceed ${cfg.max_portfolio_risk_pct}% of capital` });
      continue;
    }
    gross += pv; risk += ra; accGross += pv; accRisk += ra; accN += 1; positions += 1;
    perSector[sector] = (perSector[sector] ?? 0) + 1;
    held.add(c.symbol);
    result.push(c);
  }
  result.sort(byOrder);
  const sortedSector: Record<string, number> = {};
  for (const k of Object.keys(perSector).sort()) sortedSector[k] = perSector[k];
  return {
    results: result,
    budget: {
      capital: cfg.capital,
      open_positions: openPositions.length,
      open_gross_value: openGross / 100,
      open_risk_amount: openRisk / 100,
      accepted_positions: accN,
      accepted_gross_value: accGross / 100,
      accepted_risk_amount: accRisk / 100,
      gross_cap: grossCapP / 100,
      risk_cap: riskCapP / 100,
      per_sector: sortedSector,
    },
  };
}
